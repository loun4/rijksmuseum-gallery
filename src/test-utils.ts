export function mockLocationHref(url = 'https://domain.com') {
  Object.defineProperty(window, 'location', {
    value: {
      href: url,
    },
    writable: true,
  })
}
