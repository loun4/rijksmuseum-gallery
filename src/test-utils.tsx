/* eslint-disable react-refresh/only-export-components */
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render, RenderOptions } from '@testing-library/react'

export function mockLocationHref(url = 'https://domain.com') {
  Object.defineProperty(window, 'location', {
    value: {
      href: url,
    },
    writable: true,
  })
}

export function makeQueryClient() {
  return new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })
}

export function QueryProvider({ children }: { children: React.ReactNode }) {
  const queryClient = makeQueryClient()
  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
}

export function renderWithRQ(ui: React.ReactNode, renderOptions?: RenderOptions) {
  function Wrapper({ children }: { children: React.ReactNode }) {
    return <QueryProvider>{children}</QueryProvider>
  }

  return render(ui, { wrapper: Wrapper, ...renderOptions })
}

function getItem(index: number) {
  return {
    id: index,
    longTitle: `longTitle-${index} Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.`,
    webImage: {
      url: `image-${index}`,
    },
  }
}

export const data = [
  {
    count: 35,
    artObjects: Array.from({ length: 20 }, (_, index) => getItem(index)),
  },
  {
    count: 35,
    artObjects: Array.from({ length: 15 }, (_, index) => getItem(index + 20)),
  },
]
