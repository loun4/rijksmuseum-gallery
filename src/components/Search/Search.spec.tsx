import { cleanup, fireEvent, render } from '@testing-library/react'
import { afterEach, describe, expect, it, vi } from 'vitest'
import { Search } from './index'
import { mockLocationHref } from '@/test-utils'

describe('Search', () => {
  afterEach(cleanup)

  it('render', () => {
    const { asFragment } = render(<Search />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('populate input name from search params on first render', () => {
    mockLocationHref('https://domain.com?q=Portrait')
    const { getByDisplayValue, getByLabelText } = render(<Search />)

    expect(getByDisplayValue('Portrait')).toBeInTheDocument()
    expect(getByLabelText('Clear search')).toBeInTheDocument()
  })

  it('disable submit button and clear text icon if the input is empty', async () => {
    mockLocationHref()
    const { getByText, queryByLabelText } = render(<Search />)

    expect(getByText(/Submit/i)).toHaveAttribute('disabled')
    expect(queryByLabelText('Clear search')).toBeNull()
  })

  it('update url search params on submit', async () => {
    const replaceStateSpy = vi.spyOn(window.history, 'replaceState')
    const { getByLabelText, getByText } = render(<Search />)

    fireEvent.change(getByLabelText('Search'), { target: { value: 'Cabinet' } })
    fireEvent.click(getByText(/Submit/i))
    expect(replaceStateSpy).toHaveBeenCalledWith({}, '', 'https://domain.com/?q=Cabinet')
  })
})
