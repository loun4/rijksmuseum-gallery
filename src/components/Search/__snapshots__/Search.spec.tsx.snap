// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search > render 1`] = `
<DocumentFragment>
  <form
    class="Search"
  >
    <div
      class="Search__inputWrapper"
    >
      <input
        aria-label="Search"
        class="Search__input"
        placeholder="Search..."
        type="text"
        value=""
      />
    </div>
    <input
      class="Search__submit"
      disabled=""
      type="submit"
      value="Submit"
    />
  </form>
</DocumentFragment>
`;
